### ISSUE TITLE
Page content resizes with noticable delay

### ISSUE THREAD

**Issue by CryUshio** [electron/electron#36280]
Latest Chrome version now syncs window and render sizes.

https://user-images.githubusercontent.com/4368958/129436191-9fe2adf1-edd4-42d5-a27d-97b372dd0733.mov

Yet still, if relayout takes a lot of time, this split still occurs, especially for more complex layouts and because chrome is not optimizing for relayout performance.

_Originally posted by @0x0blu in https://github.com/electron/electron/issues/3615#issuecomment-898845463_

-------------------------------------------
#3615 #9608
Reproduces with Electron 21.x

**Comment by CryUshio** [comment]
@zcbenz It seems that resizing a page has delay in Electron.

Resizing in Chrome:
https://user-images.githubusercontent.com/30655354/200555066-3494bcc0-0689-4358-b5d9-6ef3d4b66f5f.mov

Resizing in Electron 21.2.2:
https://user-images.githubusercontent.com/30655354/200555261-ba12ecef-23b7-4392-9a7e-46d0304d2e1e.mov

**Comment by zcbenz** [comment]
Thanks for the video, yeah we should do some investigation on this.

**Comment by dtzxporter** [comment]
Any progress on the investigation?

Chrome stable has had smooth resizing for a long time, esp with the metal angle backend: `--use-angle=metal`.

Configuring the flag properly initializes metal on electron, but electron still has a massive synchronization problem when resizing the window, as apposed to chrome stable.

FWIW If anyone is debugging on `Apple Silicon`, it's probably better to profile using `metal` because on chrome stable it also eliminates overhead from the `ogl` backend.

I have been meaning to get a debug build of electron to profile myself, but I haven't had the time yet.

**Comment by bwong365** [comment]
I may have a similar issue, noticing it with Microsoft Teams and VS Code, although it really only became an issue as of today for whatever reason. I'm on Ventura 13.0.1 on an M1

https://user-images.githubusercontent.com/42486090/208151405-ba89f53e-50c6-4765-ae8d-164b4097601a.mov

**Comment by markh-discord** [comment]
I tracked this down and believe the resize issue was introduced by: https://chromium-review.googlesource.com/c/chromium/src/+/3118293.

This is clearing the `visual_properties_ack_pending_` flag prior to calling `SynchronizeVisualProperties`, which then causes the early exit to be skipped while there are prior visual properties being processed. This causes a queue of updates, which get behind and as resizing continues it gets further delayed.

This seems to affect OSX mostly, it's unclear exactly why (but I suspect) that there's simply more events being pushed.

**Comment by codebytere** [comment]
@markh-discord this issue was originally opened in 2017 - are you referring to a new different behavior?

**Comment by markh-discord** [comment]
@codebytere ,
> @markh-discord this issue was originally opened in 2017 - are you referring to a new different behavior?

Ah, so I assumed this was tracking newer behavior since I saw it was opened in November 2022, I wasn't aware it was following up on prior/older issues. Specifically though, the behavior that @bwong365 posted above is what I'm referring to on Macs - with delayed resize rendering occurring that appears to be queued up and eventually "catches" up several seconds after resizing stops.

**Comment by dtzxporter** [comment]
> I tracked this down and believe the resize issue was introduced by: https://chromium-review.googlesource.com/c/chromium/src/+/3118293.
> 
> This is clearing the `visual_properties_ack_pending_` flag prior to calling `SynchronizeVisualProperties`, which then causes the early exit to be skipped while there are prior visual properties being processed. This causes a queue of updates, which get behind and as resizing continues it gets further delayed.
> 
> This seems to affect OSX mostly, it's unclear exactly why (but I suspect) that there's simply more events being pushed.

Chrome itself doesn't have the issue though, chrome is capable of smoothly resizing the page without any issues. However the same chromium engine version in Electron is unable to do so, same flags, same everything.

It's likely related to Electron's window / chrome logic which differs from chrome.

**Comment by dtzxporter** [comment]
Not stale...

**Comment by xpf0000** [comment]
Same problem with electron23.

**Comment by bpasero** [comment]
Here is a comparison of VS Code resize performance in Electron vs Web. Notice the delay in Electron:

**Electron:**
![Recording 2023-11-23 at 09 10 01](https://github.com/electron/electron/assets/900690/94b9fd58-f5df-4229-8234-085e333e2937)

**Web:**
![Recording 2023-11-23 at 09 11 18](https://github.com/electron/electron/assets/900690/d3ee7a52-dd10-45a8-909a-1199d661e08d)

**Comment by markh-discord** [comment]
#40577 should likely resolve this.

**Comment by Andrew-web-coder** [comment]
I can see a noticeable delay using Windows.

**Comment by ostov-larion** [comment]
Same for Linux.
[Запись экрана от 2024-04-08 00-52-55.webm](https://github.com/electron/electron/assets/39136591/3db88074-967f-40cb-b9a6-67139ec418d1)

**Comment by dtzxporter** [comment]
Bump, still broken.

**Comment by devsuperking** [comment]
Bump

**Comment by FireMapleLeaves** [comment]
Bump

**Comment by EmrecanKaracayir** [comment]
Apple Silicon macOS happens on pretty much every Electron app.

**Comment by ericcastro** [comment]
pretty annoying on apple silicon as @EmrecanKaracayir said

aka an instant reminder of how electron still fails to look and feel native...

**Comment by modernve** [comment]
Bump

**Comment by nkallen** [comment]
There is a workaround that mitigates this issue significantly *on Mac only*. 

In your main process,

```
    let prev = performance.now();
    mainWindow.on('will-resize', (e: any, newBounds: Rectangle, details: WillResizeDetails) => {
        e.preventDefault();
        const now = performance.now();
        if (now - prev >= 1000 / 60) {
            prev = now;
            mainWindow.setBounds(newBounds);
        }
    });
```

This code throttles resizing to 60fps. The "jank" that is visible is due to issues in both Electron and Chrome. Chrome does not correctly double buffer the frame during resize events, and Electron has some queue load-shedding code that is poorly designed. On mac, what happens typically is that the window is resized more rapidly than the frame buffer contents are updated; and strange queuing behavior will also discard events that it shouldn't.

This fix does not resolve the issue with windows, which seems to be caused primarily by bugs in Chrome.

**Comment by bpasero** [comment]
@nkallen this indeed seems to improve the resize jank. I am still wondering though why for VS Code for example, its so much worse in Electron vs. Chrome, where the same code runs 🤔

**Comment by nkallen** [comment]
@bpasero I've reviewed the source for electron and chromium on this issue for a few hours but it's fairly complex so I still have only hypotheses.

Firstly the issue definitely still happens in Chromium https://issues.chromium.org/issues/376296999 and they are aware of it though I'm not optimistic they will fix anything anytime soon.

Secondly, the basic workflow for resizing a window should be: the resize event is received (user drags cursor), the browser performs a reflow/repaint to a buffer (this can take many milliseconds), then the operating system is instructed to resize the window and swap the buffer atomically. There may be some queuing/load shedding as well (reflow repaint takes time, new events might be received meanwhile; so the simplest thing to do is discard all but the most recent event).

I believe Chrome has implemented this code correclty on Mac but not Windows. I believe Electron has implemented this incorrectly on both, PLUS there is some very strange/incorrect queueing behavior introduced in Electron 28.

The key thing to understand is that chrome is not faster at reflow/repaint than electron, it's just not showing partial results.

**Comment by codebytere** [comment]
@nkallen  can you expand on

>  PLUS there is some very strange/incorrect queueing behavior introduced in Electron 29

**Comment by nkallen** [comment]
@codebytere On Mac only, Electron 29 introduces some type of new queuing and load-shedding behavior during resizing. The behavior persists through 33 and I have not tested more recent versions. Here is a video of the lag as observed in Electron 33. To repeat, I did a git bisect to see that it was introduced sometimes in 29 - the problem is not in Electron 28 or earlier. This specific issue (mac only) is mitigated by the `will-resize` hack mentioned previously. This is definitely a bug in Electron, not chromium. 

Unfortunately, there is also an similar but unrelated bug on windows that is even worse but it is most likely an issue in Chromium not electron. That seems to have been introduced in Electron 28.

https://github.com/user-attachments/assets/cc1d8156-8917-4c39-9ac4-33ed05daa216

**Comment by nkallen** [comment]
These are the latest versions of vscode and linear. This is on a brand new, powerful laptop, plugged in, windows 11. On both linear and vscode, there are half-written frame buffers

https://github.com/user-attachments/assets/58aadf3e-6d1a-400b-9e48-0e083a159719
https://github.com/user-attachments/assets/046e905d-c20f-47e1-996e-c9b5fff38de4


<img width="1610" alt="Image" src="https://github.com/user-attachments/assets/8732783a-4078-424f-ae69-fffc11806c30" />

**Comment by dsanders11** [comment]
@nkallen, this is a tricky issue because it's hard to objectively test it given the nature of the bug. Regarding the Windows regression between Electron 27 and 28, I've tried poking at it and I think I can see a behavior difference with a hacky test case, but I'd like to see if it also helps in your test. Can you try setting the `opacity` value for the `BrowserWindow`? Anything that's not 1.0 should work, if you set it to something like 0.99999 the window will still look fully opaque. Setting it to a non-1.0 value will effectively undo the change in #39895, which I have a sneaking suspicion might be related to the regression you're seeing.

**Comment by nkallen** [comment]
> [@nkallen](https://github.com/nkallen), this is a tricky issue because it's hard to objectively test it given the nature of the bug. Regarding the Windows regression between Electron 27 and 28, I've tried poking at it and I think I can see a behavior difference 

@dsanders11 are you able to observe the behavior in vscode and linear? the picture here https://github.com/electron/electron/issues/36280#issuecomment-2894856565 is from vscode where multiple sizes of the window are rendered to the same framebuffer simultaneously. I have two windows machines with near top-end hardware and observe this behavior on both of them, so I'm assuming it's not hard to reproduce? FWIW, going to linear's website with Chrome or Edge, the same problem occurs on windows.

**Comment by dsanders11** [comment]
> are you able to observe the behavior in vscode and linear? [...] so I'm assuming it's not hard to reproduce

If the repro case is just manually resizing the window rapidly (like in https://github.com/user-attachments/assets/046e905d-c20f-47e1-996e-c9b5fff38de4) and trying to visually notice the issue while UI elements are reflowing, I don't consider that easy to reproduce. I messed around with it on Windows for a while but most of the time could not definitively say if it was happening or not until I hacked together a test case that rapidly does the resizing on an interval and some differences became visible. So while I can somewhat observe it, it's not an objective test (e.g. something we can automate with a pass/fail).

Please try changing the `opacity` setting as I recommended in the previous comment to see if you notice a change in behavior.

**Comment by nkallen** [comment]
I don't understand. Look at the following screenshot. This is vscode. Are you not seeing this behavior in vscode on your computer? If not that's an interesting datapoint and very surprising. This is not a subtle effect that one "tries" to notice. This is very evident jank. If you make a video while moving the mouse rapidly literally every single frame will have framebuffer jank like this. Please make a video and post it?

<img width="1256" alt="Image" src="https://github.com/user-attachments/assets/b6bfbc50-298c-4d90-bf87-40b721c56d8b" />

**Comment by nkallen** [comment]
I'm sorry about my previous grumpy post. I'll test it tomorrow. It's fairly time consuming for us to test it because we need to use a fork without the pointer cage. I recommend testing it first on vscode while making a video. It should be pretty clear if you pause the video whether the problem has gone away.

**Comment by nkallen** [comment]
I think that the opacity trick completely fixes the framebuffer clobbering bug introduced in Electron v28. The improvement is huge, please see the attached video and freeze-frames. 

Thank you so much for finding this fix! I'm assuming you can find a solution where it won't be necessary to set the opacity? This satisfies my bug bounty and you earned the money. Please contact me (nick at plasticity dot xyz). This improvement will benefit vscode, linear, Plasticity, and every modern Electron app on windows.

The older bug (much older than v28), where the window resizes at a different framerate than reflow, is still there, unfortunately. This is also called out in the video. I am willing to put an additional $10k bounty for a fix for this bug. This should be fixed on both Windows and Mac.

(NOTE: all tests run on a fork of Electron 33.)

https://github.com/user-attachments/assets/15aeae50-132f-40f9-a56d-fe859d67e7b4

**Comment by dsanders11** [comment]
> Thank you so much for finding this fix! I'm assuming you can find a solution where it won't be necessary to set the opacity?

I still need to do my due diligence and confirm with a build that the issue is definitely caused by #39895 and not just a coincidence regarding transparency, but then I'll discuss with other maintainers about reverting #39895 in the short-term as the negative impact you're seeing is larger than the issue that PR aimed to fix.

Long-term this will still require more debugging/research. The issue the opacity trick fixes might be an interaction between the original bug (reflow lagging noticeably behind resizing) and that PR - when the window is marked as transparent, it may prevent the old framebuffers from being visible in the dead area (which is otherwise a black bar in your videos).

Ideally we can get some automated testing around these cases otherwise there's a high risk of regression.

**Comment by nkallen** [comment]
It does seem a challenge to write a unit test for something like this. Maybe a complicated layout where everything has a bgcolor that's a hash of the timestamp. Resizing the window and grabbing screenshots should always show a uniform color.

**Comment by billyang1222** [comment]
Electron 34，windows 10, it's even worse. Before the new area is rendered, the old content is displayed, making it look like there's an overlap on the interface. This might be an optimization of Chromium, but it looks terrible.

**Comment by nikwen** [comment]
@nkallen has hired me to work on the frame tearing issue where we see old or stale pixels on Windows.

I know how to solve it, and I have already landed some of the patches in Chromium.

This work and this post are sponsored by [Plasticity](https://www.plasticity.xyz).

### This is an upstream issue

Chrome and Edge show the same frame tearing behavior.

<img width="3248" height="2120" alt="Image" src="https://github.com/user-attachments/assets/930deff1-8294-4959-8ce7-82d438dd1548" />

### Minimal repro

The following code is enough to see visual artifacts in Electron:

```
const { app, BaseWindow } = require('electron')

app.whenReady().then(() => {
  new BaseWindow({
    backgroundColor: "red",
  })
})
```

### Workarounds

The following functions and flags mitigate the frame tearing issue:

- `app.disableHardwareAcceleration()`
- `--use-angle=warp`
- `--use-angle=vulkan`
- `--use-gl=desktop`
- `--use-gl=egl`
- `--use-gl=osmesa`
- `--use-gl=swiftshader`

The following flags are affected by the frame tearing issue:

- `--use-angle=d3d11` (this is currently the default on Windows)
- `--use-angle=gl` (falls back to Direct 3D on Windows, see `chrome://gpu/`)

None of the working flags are production ready in my opinion. They are either too slow or lack good driver support.

However, these workarounds pointed me into the right direction. They showed that the issue was in a code path that was only used with the ANGLE Direct3D backend.

I did not follow the transparency lead further. However, we'll get back to transparency later.

### What causes the issue?

Under the hood, there are two separate issues:

1. An issue where different GPU operations are out of sync, leading to previously drawn pixels being shown.
2. An issue where frames are being resized, leading to gutter that shows stale pixels.

Both of them lead to stale pixels in different situations. Both need to be fixed to fully resolve the issue.

I have fixes for both of them. One has already been merged into Chromium. The other will hopefully be merged soon.

But let's go through it slowly.

When using the ANGLE Direct 3D backend, Chromium uses the Windows [DirectComposition](https://learn.microsoft.com/en-us/windows/win32/directcomp/directcomposition-portal) API for drawing the window contents.

DirectComposition differs from all other output devices in Chromium. The DirectComposition `OutputSurface` has the capability `supports_viewporter` ([source link 1](https://source.chromium.org/chromium/chromium/src/+/main:ui/gl/dcomp_presenter.cc;l=144-146;drc=1cffe9643e1e70f1f369bbb28e015d9c5f968546), [source link 2](https://source.chromium.org/chromium/chromium/src/+/main:components/viz/service/display/output_surface.h;l=100-103;drc=c96a878b1cb45f60aac2285ffbdbd6b53dc92415)).

An output surface is a bitmap that can be drawn to, often backed by a GPU texture.

Without `supports_viewporter`, whenever the window size changes, Chromium will create a new output surface matching the new window size. Then it will draw on that surface and show it.

`supports_viewporter` tries to reduce these costly surface allocations. With `supports_viewporter`, Chromium will not allocate a new surface on every resize. Instead, it will allocate a surface that is too large for what we need to draw. Then it will only paint to and show a certain sub-rectangle (the "viewport") of that surface on screen. The other parts of the surface are not supposed to be shown on screen.

This is supposed to make resizing more efficient because all Chromium needs to do is pad the surface to the proper width and height instead of allocating a new surface on every resize. This surface resize logic lives [here](https://source.chromium.org/chromium/chromium/src/+/main:components/viz/service/display/direct_renderer.cc;l=1082;drc=c96a878b1cb45f60aac2285ffbdbd6b53dc92415).

Here's what that looks like:

<img width="2363" height="1074" alt="Image" src="https://github.com/user-attachments/assets/5b2a141b-9dd4-45d3-870f-7b836846a8c5" />

The blue rectangle is our surface. The green area is our viewport, i.e., the area of our surface that is supposed to be visible and that we actively draw to. The red rectangle is our clip rect(angle), i.e., the part of the surface that is actually being shown on screen.

As a performance optimization, only the viewport (the green area) is repainted when we get a new frame. The rest is left unchanged. This is important. We only ever repaint the green viewport. We don't update the areas outside of the viewport.

When we resize the window, what's supposed to happen is that in an atomic transaction (= at the exact same time) we repaint the viewport (= the area that's supposed to be visible on screen) and then update the clip rect to clip the surface to the new viewport size.

After the resize, it should look like this:

<img width="2363" height="1074" alt="Image" src="https://github.com/user-attachments/assets/43742f73-9256-4473-ae0b-6d9a53975c2a" />

And that's where we get to the first of our two bugs.

### First bug

Sometimes these operations can get out of sync. For example, the clip rect might get updated before the viewport is repainted. Then we get a result like this:

<img width="2363" height="1074" alt="Image" src="https://github.com/user-attachments/assets/fc888031-37ea-40a5-a713-022604f63bb7" />

We still show the old frame in the green viewport. But the clip rect has become larger and we show areas of the surface that we haven't repainted yet.

On the first resize of a window, these areas would be black. On the second resize, those areas would be filled with old pixel values. They would show whatever we have previously painted in those areas.

Similarly, in a certain edge case while making the window smaller, we might sometimes repaint the viewport before we update the clip rect.

<img width="2363" height="1074" alt="Image" src="https://github.com/user-attachments/assets/033ef5bc-dfa7-497d-9f15-6adea05416c8" />

Then parts of the clip rect might still show the previous frame because the new frame is smaller and we did not repaint any areas beyond the new viewport.

Now why do these operations not happen in sync?

We use two different Windows APIs here:

- [`IDXGISwapChain1::Present1`](https://source.chromium.org/chromium/chromium/src/+/main:ui/gl/dc_layer_tree.cc;l=1034;drc=35d26d364efb57d0386b98312ba739f7f65ae97e) — This shows the new pixels on screen / updates the new viewport.
- [`IDCompositionDevice::Commit`](https://source.chromium.org/chromium/chromium/src/+/main:gpu/command_buffer/service/shared_image/dxgi_swap_chain_image_backing.cc;l=283;drc=92e802d8fa66bafb0cca9ed32143d6148d8e0411) — This updates the clip rect.

Here's what's important to understand: both functions return synchronously on the CPU. However, what they do is they schedule tasks that run asynchronously on the GPU at a later time. Windows and its services (such as DWM) decide when they will run and in which order. So they take effect asynchronously, and not always within the same frame.

Unfortunately, Windows provides no way for us to synchronize those operations. So we need to find other approaches to fix this.

Here are the two options that I'm currently evaluating with the Chromium maintainers in [crrev.com/c/7115438](https://crrev.com/c/7115438):

1. While resizing, paint all previously drawn areas outside of the viewport transparent. This makes those areas invisible. It fixes the artifacts.
2. While resizing, switch from an `IDXGISwapChain1` to a DirectComposition surface which synchronizes updates with `IDCompositionDevice::Commit`. This also fixes the artifacts.

I'm in favor of the first option because it leads to faster resizes than the second option. It's what I've implemented in [crrev.com/c/7115438](https://crrev.com/c/7115438). But we will see which solution we will ultimately land on.

### Second bug

In addition to this first bug, there was a second bug which also led to stale pixels.

I fixed it in [crrev.com/c/7156576](https://crrev.com/c/7156576) and the fix was merged today! :tada:

Here's what was going on there:

When the user resizes the window, Chromium needs to redraw the contents of the window for the new window size. This takes some time. The new frame isn't ready immediately.

Here's a sequence of frames that demonstrates this:

<img width="3248" height="2120" alt="Image" src="https://github.com/user-attachments/assets/54bca398-8ff0-41bd-9e20-d454f75b49b0" />

<img width="3248" height="2120" alt="Image" src="https://github.com/user-attachments/assets/6624d580-f941-4ff7-a26c-65148762de74" />

<img width="3248" height="2120" alt="Image" src="https://github.com/user-attachments/assets/54c0002b-b4c9-48af-8cd8-7c35bd16751b" />

<img width="3248" height="2120" alt="Image" src="https://github.com/user-attachments/assets/3edae2e1-8449-40d3-8614-91f906e82028" />

<img width="3248" height="2120" alt="Image" src="https://github.com/user-attachments/assets/9cd322f8-d047-4949-b089-fce31b3a0698" />

At a certain time during the resize, Windows tells us: "The window is 1,000 pixels wide." But the frames that the browser compositor produces are lagging behind. The last frame we have painted might be 600 pixels wide.

Historically, Chromium used to skip frames where the width of the window did not match the width of the frame that it last painted. It would decide to just not update the window.

However, that would often lead to the window contents not being updated at all until the resize operation was finished.

So [in 2015](https://crrev.com/1426453006) someone decided: "Why not show these frames? They might not match the window size perfectly, but at least we can show _something_."

It would lead to gutter, but at that time the gutter was black. So that was better than the previous implementation.

Now, 10 years later with DirectComposition, that gutter is often filled with stale pixels.

Let's look at what's happening there:

Every frame consists of multiple render passes. The render passes represent the various things that should be drawn on screen. From complicated bitmaps to rectangles filled with solid colors.

Every frame has a root render pass, which contains all other render passes and glues them together. (Render passes are arranged in a tree structure and the root render pass is the root of that tree.)

So now during a resize, we'd get to a point where we know the window is 1,000 pixels wide. Accordingly, we'd adjust the viewport of our output surface to also be 1,000 pixels wide. But the frame that we just received is only 600 pixels wide.

The [optimization from 2015](https://source.chromium.org/chromium/chromium/src/+/main:components/viz/service/display/display.cc;l=1053-1062;drc=152a83bb9fcba4ea3a7b9f8efbe0f6204f95eadd) would then go and change the width of the root render pass to also be 1,000 pixels. But it wouldn't change what the render passes would actually draw on screen. They'd still only contain instructions to draw a picture that is 600 pixels wide.

Here's what that would look like:

<img width="2363" height="1074" alt="Image" src="https://github.com/user-attachments/assets/3dfd12a8-33e1-4327-9820-22bfd5eea72f" />

The yellow area is the area in which the render passes of the frame actually drew something. It's 600 pixels wide.

However, our green viewport and our red clip rect are 1,000 pixels wide. That's the area that we show on screen.

After all, the width attribute of the root render pass claimed that it would redraw the full area of 1,000 pixels.

But because we had no draw instructions for the 400 pixels on the right, those areas didn't get updated.

On the first resize, we'd show black pixels there. (That's the color that we initialize the surface with.)

On subsequent resizes, those areas would show whatever was drawn to them before. We'd see stale pixels.

As mentioned before, we merged a [fix](https://crrev.com/c/7156576) for this issue into Chromium today.

The fix changes what we do when we receive a frame with a different size than our window. Instead of resizing the frame and adding gutter that contains stale pixels, we [resize our viewport](https://source.chromium.org/chromium/chromium/src/+/main:components/viz/service/frame_sinks/root_compositor_frame_sink_impl.cc;l=555-556;drc=78bd63326c0a2489e6b016140aa4d84014af077a) and our clip rect.

<img width="2363" height="1074" alt="Image" src="https://github.com/user-attachments/assets/24768a15-139b-44f7-80c3-62669ee1d711" />

We clip our surface to the size of the frame that we received. We don't show anything beyond the 600 pixels that we have draw instructions for.

Voila, no more gutter, no more stale pixels!

(Side note: Without `supports_viewporter`, this would be an expensive operation because it would allocate a new output surface. However, with DirectComposition, we use the "viewporter" feature. So we don't reallocate the surface when we change the viewport size. We just make a different portion of it visible. Thus, it is a cheap operation.)

### Why did the transparency workaround help?

`skia_renderer.cc` contains the following [lines of code](https://source.chromium.org/chromium/chromium/src/+/main:components/viz/service/display/skia_renderer.cc;l=1299-1301;drc=28379c36f1f54580b2a51157090cf02333ab12d9):

```cpp
if (current_frame()->current_render_pass->has_transparent_background) {
  ClearCanvas(SkColors::kTransparent);
}
```

If the background of whatever was being painted was partially or fully transparent, the code would fill the viewport with a transparent color before drawing the render passes on the surface.

This would eliminate most (but not all) of the stale pixels. It helped whenever the stale pixels were within the viewport. It did not help when we'd show stale pixels that were outside of the viewport.

### Next steps

My fix for the second bug landed in Chromium today via [crrev.com/c/7156576](https://crrev.com/c/7156576). On my machine, fixing the second bug removes about half of the stale pixels. I'll backport the fix to Electron within the next few days.

I also have a fix for the first bug up for review in [crrev.com/c/7115438](https://crrev.com/c/7115438). It works and fully removes the remaining stale pixels. Right now, it's mostly a matter of communicating with the Chromium maintainers to make sure we don't break any non-Window or non-DirectComposition use cases and to choose the best of all available solutions. Once a fix is merged, I will backport it to Electron as well.

### Thanks

Big thanks to @nkallen and [Plasticity](https://www.plasticity.xyz) for funding this work! Also, thanks to Michael Tang and Vasiliy Telezhnikov from the Chromium team for their help.

### Last words

This has been the hardest bug I have ever worked on. (And I have worked on many hard bugs in 18 years of writing software.) But it has also been one of the most fun projects I have ever worked on.

There is more nuance to it than I could include in this write-up. I had to leave out certain details for the sake of simplicity.

**Comment by nikwen** [comment]
### Update

**All patches have been merged into Chromium! ([#1](https://chromium-review.googlesource.com/c/chromium/src/+/7129658), [#2](https://chromium-review.googlesource.com/c/chromium/src/+/7156576), [#3](https://chromium-review.googlesource.com/c/chromium/src/+/7210913), [#4](https://chromium-review.googlesource.com/c/chromium/src/+/7115438)) 🎉**

One of them has been [merged](https://github.com/electron/electron/pull/49076) into Electron 39 (but is not part of any release yet).  
I have a [pull request](https://github.com/electron/electron/pull/49138) open to backport the other patches to Electron 39, too. (Reviews welcome!)  
The patches will make it into Electron v40 and v41 through one of the next Chromium rolls.

The patches fully fix the [issue](https://github.com/electron/electron/issues/36280#issuecomment-2894856565) where [old pixels](https://github.com/electron/electron/issues/36280#issuecomment-2934365927) would show up during window resizes on Windows (highlighted in red):

<img width="1901" height="1422" alt="Old frames being shown while resizing window on Windows" src="https://github.com/user-attachments/assets/a7de0851-d6c2-41a2-8cdf-2425f6aa2347" />

### Where to take it from here

The following issues still require investigation:

#### 1. Windows + macOS: In Electron, web content [resizes slower](https://github.com/electron/electron/issues/36280#issue-1439824959) than in Chromium.

I'd assume that this issue is easier to solve than the second issue below.

It might be caused by differences between Electron's window resize code and Chromium's window resize code. There's a good chance its root is in the Electron source code, not in Chromium.

It also has a higher impact than the second issue.

#### 2. Windows only: In both Electron and Chromium, the size of the compositor frames (the size of the drawn website) lags behind the window size.

This issue needs to be fixed upstream in Chromium. It might require Windows-specific optimizations of the compositor (`cc`) code and maybe also optimizations in the display compositor (`viz`). That could be a large project.

We should start by reporting it to Chromium upstream. I haven't see an issue for it in their bug tracker.

**Comment by nikwen** [comment]
If you want to test the patches, here is a [Windows x64 release build](https://drive.google.com/file/d/1FPxZSxyAlVZsOwDvbwOXjAO660essu8o/view) made from commit [`9947a0133c70bdf8eb95771274d7677e34321753`](https://github.com/electron/electron/commit/9947a0133c70bdf8eb95771274d7677e34321753) on the [`nikwen/backport-resize-fixes/39-x-y`](https://github.com/electron/electron/tree/nikwen/backport-resize-fixes/39-x-y) branch.

SHA256SUM: `845e3ea89382686985248d472d475c02a4a70c32366dea4260c9949ef65cf5a6`

**Comment by nikwen** [comment]
The patches are part of [Electron 39.2.6](https://releases.electronjs.org/release/v39.2.6) now!

I noticed that the last two upstream patches just barely missed the deadline for Chromium 144 (which Electron 40 is based on), so I'll backport the patches to Electron 40, too. (And then I'll also do `main`, because why not?)

**Comment by nikwen** [comment]
Two separate bugs were reported in this GitHub issue:

1. [Electron resizes slower than Chromium](https://github.com/electron/electron/issues/36280#issuecomment-1824041120). ❌ Not fixed yet.
2. [Visual artifacts while resizing on Windows](https://github.com/electron/electron/issues/36280#issuecomment-2894856565). ✅ Fixed on Electron 39+.

I will not get to work on the first one immediately. So I'll unassign myself from this issue for now.
